rust/frontend/src/value.rs

```diff
@@ -0,0 +1,540 @@
+//! This module implements [`TVMArgValue`] and [`TVMRetValue`] types and their conversions
+//! to other support TVMValue.
+//!
+//! # Example
+//!
+//! ```
+//! let a = 42i8;
+//! let b = TVMArgValue::from(&a);
+//! assert_eq!(b.to_int() as i8, a);
+//! ```
+
+use std::ffi::{CStr, CString};
+use std::fmt::{self, Debug, Formatter};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::os::raw::{c_char, c_void};
+
+use ts;
+
+use ty::TypeCode;
+use Function;
+use Module;
+use NDArray;
+use TVMByteArray;
+use TVMContext;
+use TVMDeviceType;
+use TVMType;
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub(crate) enum ValueKind {
+    Int,
+    Float,
+    Handle,
+    Str,
+    Bytes,
+    Type,
+    Context,
+    Return,
+}
+
+/// Wrapper around underlying TVMValue.
+#[derive(Clone)]
+pub struct TVMValue {
+    pub(crate) kind: ValueKind,
+    pub(crate) inner: ts::TVMValue,
+}
+
+impl TVMValue {
+    pub(crate) fn new(kind: ValueKind, inner: ts::TVMValue) -> Self {
+        TVMValue { kind, inner }
+    }
+
+    pub fn to_int(&self) -> i64 {
+        unsafe { self.inner.v_int64 }
+    }
+}
+
+macro_rules! impl_prim_val {
+    ($type:ty, $kind:expr, $field:ident, $cast:ty) => {
+        impl<'a> From<&'a $type> for TVMValue {
+            fn from(arg: &$type) -> Self {
+                let inner = ts::TVMValue {
+                    $field: *arg as $cast,
+                };
+                Self::new($kind, inner)
+            }
+        }
+
+        impl<'a> From<&'a mut $type> for TVMValue {
+            fn from(arg: &mut $type) -> Self {
+                let inner = ts::TVMValue {
+                    $field: *arg as $cast,
+                };
+                Self::new($kind, inner)
+            }
+        }
+    };
+}
+
+impl_prim_val!(usize, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i64, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i32, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i16, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i8, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u64, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u32, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u16, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u8, ValueKind::Int, v_int64, i64);
+impl_prim_val!(bool, ValueKind::Int, v_int64, i64);
+impl_prim_val!(f64, ValueKind::Float, v_float64, f64);
+impl_prim_val!(f32, ValueKind::Float, v_float64, f64);
+
+impl<'a> From<&'a [u8]> for TVMValue {
+    fn from(arg: &[u8]) -> TVMValue {
+        let len = arg.len();
+        let arg = CString::new(arg).unwrap();
+        let arr = ts::TVMByteArray {
+            data: arg.as_ptr() as *mut c_char,
+            size: len,
+        };
+        let inner = ts::TVMValue {
+            v_handle: &arr as *const _ as *mut c_void,
+        };
+        mem::forget(arg);
+        Self::new(ValueKind::Handle, inner)
+    }
+}
+
+impl<'a> From<&'a mut [u8]> for TVMValue {
```

Comment:
please replace these copy+paste functions with a macro

Reply:
I wanted to actually, but there're not copy past and each (with exception of one or two) have different pointer conversions.

Reply:
okay but at the very least, can you refactor it so that the from `&` and from `&mut` use a common implementation?

Reply:
sure, I'll refactor.